src/webui/www/tstool.py

```diff
@@ -0,0 +1,192 @@
+
+    # add messages for new contexts
+    for context_name in sources:
+        context = ET.SubElement(root, 'context')
+        ET.SubElement(context, 'name').text = context_name
+
+        for source in sources[context_name]:
+            message = ET.SubElement(context, 'message')
+            ET.SubElement(message, 'source').text = source
+            ET.SubElement(message, 'translation', {'type': 'unfinished'}) 
+
+    # prettify output xml
+    indent = ' ' * 4
+    root.text = '\n'
+    for context in root.findall('./context'):
+        context.text = '\n' + indent
+        context.tail = '\n'
+        context.find('./name').tail = '\n' + indent
+        messages = context.findall('./message')
+        if len(messages) == 0: continue
+        
+        for message in messages:
+            message.text = '\n' + (indent * 2)
+            message.tail = '\n' + indent
+            elems = message.findall('./')
+            if len(elems) == 0: continue
+            
+            for elem in elems:
+                elem.tail = '\n' + (indent * 2)
+            elems[-1:][0].tail = '\n' + indent
+        messages[-1:][0].tail = '\n'
+
+    try:
+        file = open(filename, mode = 'wb')
+        file.write(b'<?xml version="1.0" encoding="utf-8"?>\n'
+                   b'<!DOCTYPE TS>\n')
+        tree.write(file, encoding = 'utf-8')
+    except:
+        print('\tFailed to write %s!' % (normpath(filename)))
+
+def main():
+    global no_obsolete, www_folder, ts_folder
+    
+    argp = argparse.ArgumentParser(
+        prog = 'tstool.py', description = 'Update qBittorrent WebUI translation files.')
+    argp.add_argument('--no-obsolete', dest = 'no_obsolete', action = 'store_true',
+                      default = no_obsolete,
+                      help = 'remove obsolete messages (default: mark them as obsolete)')
+    argp.add_argument('--www-folder', dest = 'www_folder', action = 'store',
+                      default = www_folder,
+                      help = 'folder with WebUI source files (default: "%s")' % (www_folder))
+    argp.add_argument('--ts-folder', dest = 'ts_folder', action = 'store',
+                      default = ts_folder,
+                      help = 'folder with WebUI translation files (default: "%s")' % (ts_folder))
+
+    args = argp.parse_args()
+    no_obsolete = args.no_obsolete
+    www_folder = args.www_folder
+    ts_folder = args.ts_folder
+    
+    print("Processing source files... ")
+    nfiles = 0
+    for root, dirs, files in os.walk(www_folder):
+        for file in files:
+            if splitext(file)[-1] in accepted_exts:
+                parseSource(join(root, file))
+                nfiles += 1
+
+    if nfiles == 0:
+        print("No source files found!")
+        return
+
+    nstrings = sum(len(sublist) for sublist in source_ts)
+    print("Found %d strings within %d contexts." % (nstrings, len(source_ts)))
+    print("")
+    
+    print("Processing translation files... ")
```

Comment:
nitpick, redundant space at end of string

Reply:
PR updated!